### ISSUE TITLE
Enhance Debian and Ubuntu support

### ISSUE THREAD

**Issue by paususe** [openSUSE/mentoring#127]
**Project Title:** Enhance Debian and Ubuntu support

**Description:** Uyuni currently supports Debian and Ubuntu officially as Salt clients but some bits are missing. This project is about implementing what's missing.

**Deliverable:** As many as possible of this list

1. Generate and process Ubuntu patches and security information (CVE information from the Ubuntu Security Notices - https://usn.ubuntu.com/). This is mostly backend work in Python (maybe it can even be implemented in pure Salt but IMHO the code will be harder to maintain).
  An alternative approach is to process the mails sent to the ubuntu-security-announce mailing list, like https://github.com/Twhitehouse-SSHIS/spacewalk-ubuntu-errata does but my feeling is this is not complete. Also, those scripts should be made a bit more flexible and integrated with Uyuni (e. g. instead of hardcoding what version of Ubuntu to download information for, get the list of versions from the enabled products in Uyuni, which is rather easy)
2. Generate and process Debian patches and security information (CVE information from the Debian Security Advisories - https://security-tracker.debian.org/tracker/, it's available in JSON format from https://security-tracker.debian.org/tracker/data/json). This is mostly backend work in Python (maybe it can even be implemented in pure Salt but IMHO the code will be harder to maintain).
  An alternative approach is to process the mails sent to the debian-security-announce mailing list, like https://github.com/philicious/spacewalk-scripts does but my feeling is this is not complete. Also, those scripts should be made a bit more flexible and integrated with Uyuni (e. g. instead of hardcoding what version of Debian to download information for, get the list of versions from the enabled products in Uyuni, which is rather easy)
3. Autoinstallation of Ubuntu. Depending on the version of Ubuntu to be installed, this requires implementing either preseed or kickseed (until Ubuntu 18.04) or [cloud-init](https://ubuntu.com/server/docs/install/autoinstall-quickstart) and/or [curtin](https://curtin.readthedocs.io/en/latest/topics/storage.html) (from 20.04; this is where I would start).
4. Autoinstallation of Debian. This requires implementing [preseed](https://wiki.debian.org/DebianInstaller/Preseed) support, which should be mostly backend work in Java.
5. Add Debian/Ubuntu-specific Content Lifecycle Management filters, or enhance the existing ones, with Debian/Ubuntu specifics.
6. Allow building Debian/Ubuntu OS images (mostly a matter of providing a Kiwi template, requires some very small enablement in Java) and/or containers
7. Implement onboarding of Ubuntu clients without requiring manual enablement: https://github.com/uyuni-project/uyuni-rfc/pull/30. This requires implementing a Salt state to be run [when bootstrapping a client](https://github.com/uyuni-project/uyuni/tree/master/susemanager-utils/susemanager-sls/salt/bootstrap) and it will also be useful for Alibaba Cloud Linux 2.

**Mentor:** @paususe + another Uyuni developer TBD from SUSE

**Skills:** You will need to have a good understanding of Linux, and be comfortable with version control using git and GitHub. You will need some PostgreSQL, JSON, Python and Java skills. Familiarity with Debian and package management.

**Skill Level:** Medium

**Get started:**
* Uyuni development in no time: https://github.com/uyuni-project/uyuni/wiki/Uyuni-development-in-no-time
* Getting started hacking Uyuni: http://bosdonnat.fr/slides/openSUSEAsiaSummit19/#/about
* Download Ubuntu 20.04 and install it in a virtual machine
* Download openSUSE Leap 15.2 and [install Uyuni Server](https://www.uyuni-project.org/pages/stable-version.html#serversetup)
* Add the Ubuntu (yes, Ubuntu, not Debian!) repositories with spacewalk-common-channels, sync them and create a bootstrap repository. Onboard an Ubuntu client.
* You will notice the "Patches" sidebar in Uyuni does not show any information about security patches.
* If you install a package on the Ubuntu client from the Uyuni Server, the WebUI will show the right information; if you install a paackge on the Ubuntu client by ssh-ing into the Ubuntu client and running apt, then the WebUI will not show that package as installed (because of the missing aptnotify beacon).
* Introduction to Uyuni and systems management (presentation): https://github.com/uyuni-project/uyuni/wiki/Presentations
* Uyuni wiki: https://github.com/uyuni-project/uyuni/wiki
* Check these links for examples of OS-enablement code:
  * Alibaba Cloud Linux 2 (complete example, includes everything) : https://github.com/uyuni-project/uyuni/commit/7d3ba0c7f9c914ac71afe2590a096f459437908a
  * Amazon Linux 2 (product identification and entitlement management is missing) : https://github.com/uyuni-project/uyuni/commit/7ac753910cff7a84d172df35bf34fb15c14fa3b4
  * Alma Linux 8 (product identification and entitlement management is missing) : https://github.com/uyuni-project/uyuni/commit/c5375e3f974fa3a8aa27ce9c2bbbd553ac12a57e

**Comment by ganpa3** [comment]
> Debian is not officially supported.

The [wiki](https://github.com/uyuni-project/uyuni/wiki/Next-candidate-operating-systems#intro) mentions that Debian 9 and 10 are supported?
So, is this project still valid?

**Comment by paususe** [comment]
> > Debian is not officially supported.
> 
> The [wiki](https://github.com/uyuni-project/uyuni/wiki/Next-candidate-operating-systems#intro) mentions that Debian 9 and 10 are supported?
> So, is this project still valid?

Yes, the project is still valid. I will modify the description with the missing stuff. 

TL;DR: currently, the most important items would be errata information and autoinstallation (which are enough for a GSoC project).

**Comment by paususe** [comment]
@ganpa3 I have updated the description. Hopefully, it is more useful now. You do not need to do all of that, only some parts. I would start with e. g. 1 and 2 (errata information), which are easier.
